File: matmul/mpi/solution/mpi_mm_v4.f90
module mat_param

  implicit none

  integer :: nra                   ! Number of rows for matrix A.
  integer :: nca                   ! Number of columns for matrix A.
  integer :: ncb                   ! Number of rows for matrix B. 

  real*8,allocatable :: a(:,:)     ! Local matrix A
  real*8,allocatable :: b(:,:)     ! Local matrix B
  real*8,allocatable :: c(:,:)     ! Local matrix C
  real*8,allocatable :: cfull(:,:) ! Full matrix C

  integer :: nsrow                 ! Number of rows in the sub-matrix.
  integer :: nscol                 ! Number of colums in the sub-matrix.

end module mat_param

module mpi_param
  
  use mpi

  implicit none

  integer :: nprocs                ! Number of processes
  integer :: myrank                ! Rank of current process
  integer :: mpi_err               ! MPI error flag
  integer :: errcode               ! MPI error code

  integer :: status(MPI_STATUS_SIZE)

  integer :: vtype                 ! MPI vector data type. 

  integer :: nprows                ! Number of rows in the process grid.
  integer :: npcols                ! Number of colums in the process grid.

  integer :: iprow                 ! Row index of current process in the process grid. 
  integer :: ipcol                 ! Column index of current process in the process grid.

end module mpi_param

program ser_matmul

  use mat_param
  use mpi_param

  implicit none

  integer :: argc                  ! Number of command line arguments.
  integer :: iargc                 ! External function returning argc.
  character*80 :: arg              ! Container for command line arguments.

  integer :: i,j,k                 ! Scratch variables.

  integer :: ipeek,jpeek           ! Indices for the element used for result validation.
  integer :: ispeek,jspeek

  integer*8, dimension(8) :: value ! Time variable.

  real*8 :: flops                  ! Number of floating point operations needed for the matrix multiplication.
  real*8 :: init_time              ! Time variable. 
  real*8 :: start_time             ! Time variable.
  real*8 :: end_time               ! Time variable.

  ! Initialize MPI environment.

  call mpi_init(mpi_err)

  ! Find out the total number of processes and the rank of current process.

  call mpi_comm_rank(mpi_comm_world, myrank, mpi_err)
  call mpi_comm_size(mpi_comm_world, nprocs, mpi_err)
  
  ! Read and validate command line parameters.

  argc=iargc()
  if (argc .ne. 5) then

     ! Incorrect number of arguments. Print the usage info.

     if (myrank .eq. 0) then
        write(*,*) "Usage: ser_mm nrows nprow npcol irow icol"
     endif
     call mpi_abort(mpi_comm_world, errcode, mpi_err)
     stop

  else

     ! Process the arguments.

     call getarg(1, arg)
     read(arg, '(i)') nra
     call getarg(2, arg)
     read(arg, '(i)') nprows
     call getarg(3, arg)
     read(arg, '(i)') npcols
     call getarg(4, arg)
     read(arg, '(i)') ipeek
     call getarg(5, arg)
     read(arg, '(i)') jpeek

  endif

  ! We only deal with square matrices in this exercise.

  nca=nra
  ncb=nra

  ! Perform some sanity checks.

  if (nprocs .ne. nprows*npcols) then
     if (myrank .eq. 0) then
        write(*,*) "The nubmer of processes is not equal to nprows*npcols!"
     endif
     call mpi_abort(mpi_comm_world, errcode, mpi_err)
     stop
  endif

  if (mod(nra,nprows) .ne. 0) then
     if (myrank .eq. 0) then
        write(*,*) "NRA is not a multiple of nprows!"
     endif
     call mpi_abort(mpi_comm_world, errcode, mpi_err)
     stop
  endif

  if (mod(ncb,npcols) .ne. 0) then
     if (myrank .eq. 0) then
        write(*,*) "NCB is not a multiple of npcols!"
     endif
     call mpi_abort(mpi_comm_world, errcode, mpi_err)
     stop   
  endif

  ! Calculate the number of floating point operations needed for the matrix multiplication

  flops = 2d0 * float(nra) * float(nca) * float(ncb)

  ! Find out current time.

  call date_and_time(VALUES=value)
  init_time = float(value(6)*60) + float(value(7)) + float(value(8))/1000d0

  ! Find out the row and column indcies of the current process in the process grid.
  ! It does not matter whether the index starts from 0 or 1 as long as the bookkeeping is tidy.
  ! We choose 0 in this example.

  iprow=myrank/npcols
  ipcol=mod(myrank,npcols)

  ! Calculate the dimensions of the sub-matrix.

  nsrow=nra/nprows
  nscol=ncb/npcols

  ! Allocate memory for sub-A, sub-B and sub-C.

  allocate(a(nsrow,nca), &
           b(nca,nscol), &
           c(nsrow,nscol))

  ! Allocate full matrix C at the root process.

  if (myrank.eq.0) then
     allocate(cfull(nra,ncb))
  endif

  ! Initialize sub-A, sub-B and sub-C.

  do j=1,nca
     do i=1,nsrow
        a(i,j)=iprow*nsrow+i+j
     enddo
  enddo
  do j=1,nscol
     do i=1,nca
        b(i,j)=i*(ipcol*nscol+j)
     enddo
  enddo
  c=0.d0

  ! Find out current time.

  call date_and_time(VALUES=value)
  start_time = float(value(6)*60) + float(value(7)) + float(value(8))/1000d0

  ! Matrix multiplication.

  do i=1,nsrow
     do j=1,nscol
        do k=1,nca
           c(i,j)=c(i,j)+a(i,k)*b(k,j)
        enddo
     enddo
  enddo

  ! Define a vector data type.

  call mpi_type_vector(nscol,nsrow,nra,mpi_real8,vtype,mpi_err)
  call mpi_type_commit(vtype,mpi_err)

  ! Send the results to the root process.

  if (myrank.eq.0) then

     ! Move the local copy.
     cfull(1:nsrow,1:nscol)=c

     ! Receive from other processes.
     do i=1,nprocs-1
        call mpi_recv(cfull(i/npcols*nsrow+1,mod(i,npcols)*nscol+1),1,vtype,i,1,mpi_comm_world,status,mpi_err)
     enddo

  else

     call mpi_send(c,nsrow*nscol,mpi_real8,0,1,mpi_comm_world,mpi_err)

  endif

  call mpi_type_free(vtype,mpi_err)

  ! Find out current time.

  call date_and_time(VALUES=value)
  end_time = float(value(6)*60) + float(value(7)) + float(value(8))/1000d0

  ! Print timing information to screen.

  if (myrank .eq. 0) then

     write(*,*)
     write(*,'(A,T12,f7.3)') 'Init Time: ',start_time - init_time
     write(*,'(A,T12,f7.3)') 'Calc Time: ',end_time - start_time
     write(*,'(A,T12,f7.3)') 'GFlops: ',1d-9 * flops/(end_time - start_time)
     write(*,*)

  endif

  ! Validate the result.

  if (myrank .eq. 0) then

     write(*,'(A)') 'Value calculated by the program:'
     write(*,'(A,es14.6)') 'C(i,j) =',cfull(ipeek,jpeek)
     write(*,'(A)') 'It should be:'
     write(*,'(A,es14.6)') 'C(i,j) =',float(ipeek)*float(jpeek)*float(nra)*(float(nra)+1)/2 + &
          float(jpeek)*float(nra)*(float(nra)+1)*(2*float(nra)+1)/6
     write(*,*)

  endif

  ! Deallocate A, B and C.
  deallocate(a,b,c)
  if (myrank.eq.0) then
     deallocate(cfull)
  endif
  
  call mpi_finalize(mpi_err)

  stop

end program ser_matmul

File: matmul/mpi/solution/mpi_mm_v4.c
#include "mpi.h"
#include <stdio.h>
#include <stdlib.h>

typedef double real;

int main (int argc, char *argv[])
{
    int nra,
        nca,
        ncb,
        nprocs,
        myrank,
        mpi_err=99,
        errcode;
    
    int i,j,k;
    int ipeek,jpeek;
    int nprows,npcols;

    int nsrow,nscol,iprow,ipcol;
    
    real flops,init_time,start_time,end_time;
    
    MPI_Status status;
    MPI_Datatype vtype;

    real** a;
    real** b;
    real** c;
    real** cfull;
    
    MPI_Init(&argc,&argv);
    MPI_Comm_rank(MPI_COMM_WORLD,&myrank);
    MPI_Comm_size(MPI_COMM_WORLD,&nprocs);
   
    printf("argc=%d\n",argc); 
    printf("myrank=%d\n",myrank); 
    if (6!=argc) {
        if (0==myrank)
            printf("Usage: ser_mm nrows nprow npcol irow icol\n");
        MPI_Abort(MPI_COMM_WORLD, mpi_err);
        exit(1);
    }
    else {
        nra=atoi(argv[1]);
        nprows=atoi(argv[2]);
        npcols=atoi(argv[3]);
        ipeek=atoi(argv[4]);
        jpeek=atoi(argv[5]);
        printf("%d,%d,%d,%d,%d\n",nra,nprows,npcols,ipeek,jpeek);
    }
    
    nca=nra, ncb=nra;
    
    if (nprocs!=nprows*npcols){
        if (0!=myrank){
            printf("The nubmer of processes is not equal to nprows*npcols!\n");
        }
        MPI_Abort(MPI_COMM_WORLD, mpi_err);
        exit(1);
    }
   
    if (0!=(nra % nprows)) {
        if (0==myrank)
           printf("Nrows is not a multiple of nprocs!\n");
        MPI_Abort(MPI_COMM_WORLD, mpi_err);
        exit(1);
    }
    
    if (0!=(ncb % npcols)) {
        if (0==myrank)
           printf("Nrows is not a multiple of nprocs!\n");
        MPI_Abort(MPI_COMM_WORLD, mpi_err);
        exit(1);
    }
//    
    flops = 2.0*nra*nca*ncb;
    init_time = MPI_Wtime();
    
    iprow= myrank/npcols;
    ipcol= myrank%npcols;
    
    nsrow=nra/nprows;
    nscol=ncb/npcols;

    // allocate memory space for A, B and C
    a=malloc(nsrow*sizeof(real*));
    for (i=0;i<nsrow;i++){
        a[i]=malloc(nca*sizeof(real));
    }
    b=malloc(nca*sizeof(real*));
    for (i=0;i<nca;i++){
        b[i]=malloc(nscol*sizeof(real));
    }
    c=malloc(nsrow*sizeof(real*));
    for (i=0;i<nscol;i++){
        c[i]=malloc(nscol*sizeof(real));
    }
    
    if (0==myrank) {
        cfull=malloc(nra*sizeof(real*));
        for (i=0;i<nra;i++)
            cfull[i]=malloc(ncb*sizeof(real));
    }

    // Initialize sub-A, sub-B and sub-C.
    for (i=0;i<nsrow;i++)
        for (j=0;j<nca;j++)
            a[i][j]=iprow*nsrow+i+j;
    
    for (i=0;i<nca;i++)
        for (j=0;j<nscol;j++)
            b[i][j]=i*(ipcol*nscol+j);
            
    for (i=0;i<nsrow;i++)
        for (j=0;j<nscol;j++)
            c[i][j]=0.0;    
    
    // find out current time
    start_time = MPI_Wtime();
    
    for (i=0;i<nsrow;i++)
        for (j=0;j<nscol;j++)
            for (k=0;k<nca;k++)
                c[i][j] += a[i][k]*b[k][j];

    MPI_Type_vector(nsrow,nscol,ncb,MPI_DOUBLE,&vtype);
    MPI_Type_commit(&vtype);
    
    if (0==myrank){
        for (i=0;i<nsrow;i++)
            for (j=0;j<nscol;j++)
                cfull[i][j]=c[i][j];
        for (i=1;i<=nprocs-1;i++)
            ;//MPI_Recv(cfull(i/npcols*nsrow+1,(i%npcols)*nscol+1),1,vtype,i,1,mpi_comm_world,status);
    }
    else {
        ;//MPI_Send(c,nsrow*nscol,MPI_DOUBLE,0,1,MPI_COMM_WORLD);
    }
    
    MPI_Type_free(&vtype);
    // find out current time
    end_time = MPI_Wtime();
 
   if (0==myrank) {
       printf("Init Time: %7.3f\n", start_time - init_time);
       printf("Calc Time: %7.3f\n", end_time - start_time);
       printf("GFlops: %7.3f\n", 1.0e-9* flops/(end_time - start_time));
   }

   if (0==myrank) {
       printf("Value calculated by the program:\n");
       printf("c[i][j]=%14.6e\n",cfull[ipeek][jpeek]);
       printf("It should be:\n");
       printf("c[i][j]=%14.6e\n",(float)(ipeek)*(float)(jpeek)*(float)(nra)*((float)nra-1.0)/2.0
                               + (float)(jpeek)*(float)(nra)*((float)(nra)-1.0)*(2.0*(float)(nra)-1.0)/6.0);
    }
    
    // free A-C
    for (i=0;i<nsrow;i++){
        free(a[i]);
    }
    free(a);
    for (i=0;i<nca;i++){
        free(b[i]);
    }
    free(b);
    for (i=0;i<nsrow;i++){
        free(c[i]);
    }
    free(c);
    if (0==myrank) {
        for (i=0;i<nra;i++)
            free(cfull[i]);
        free(cfull);
    }

    MPI_Finalize(); 
}
